src/itemstackmetadata.cpp

```diff
@@ -58,15 +59,28 @@ bool ItemStackMetadata::setString(const std::string &name, const std::string &va
 	return result;
 }
 
-void ItemStackMetadata::serialize(std::ostream &os) const
+void ItemStackMetadata::serialize(std::ostream &os, bool disk) const
 {
 	std::ostringstream os2;
 	os2 << DESERIALIZE_START;
+	std::string unsent_fields;
 	for (const auto &stringvar : m_stringvars) {
+		if (! disk
+				&& stringvar.first != TOOLCAP_KEY
+				&& stringvar.first != "description"
+				&& stringvar.first != "color"
+				&& stringvar.first != "short_description"
+				&& stringvar.first != "palette_index") {
+			unsent_fields += stringvar.first + "=" + stringvar.second;
```

Comment:
`unsent_fields.append()` is faster than string concat.

Or even, make it an `std::ostringsteam` and re-use the code below like so:
```C++
//std::ostringsteam os_hash;

bool silent_output = !disk && .. && .. && ..;
if (!stringvar.first.empty() || !stringvar.second.empty())
	(silent_output ? os_hash : os2) << stringvar.first << DESERIALIZE_KV_DELIM
		<< stringvar.second << DESERIALIZE_PAIR_DELIM;

```

Reply:
Why can't unsent_fields just be a boolean?

Reply:
There must be a hash, otherwise the client prediction will not work as intended. It would assume the stacks are identical and allows to merge them.

Reply:
Done.